src/api/app/models/concerns/has_relationships.rb

```diff
@@ -37,6 +37,10 @@ def render_relationships(xml)
     end
   end
 
+  def local_roles_for_user(user)
+    relationships.where(id: Relationship.where(user: user).or(Relationship.where(group: user.group_ids)), role: Role.local_roles).extract_associated(:role)
```

Comment:
Instead of fetching `Relationship` and then asking for its roles, could we query for `Role` directly?

Reply:
> Instead of fetching `Relationship` and then asking for its roles, could we query for `Role` directly?

Yes. We could do:

```ruby
Role.joins(:relationships).where(
  id: Role.local_roles,
  relationships: relationships.where(id: Relationship.where(user: user).or(Relationship.where(group: user.group_ids)))
).distinct
```

But first, we should change this line:

https://github.com/openSUSE/open-build-service/blob/90db6c4a1ab2f02c0396b10613b360923594192b/src/api/app/models/role.rb#L26

to something similar to:

```ruby
has_many :relationships, dependent: :destroy
```

I'm not sure we should... :grimacing:

Reply:
I created these PR's. The second PR replaces this PR. Please take a look:
- #16300
- #16301